cpp/include/cudf/join/sort_merge_join.hpp

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/export.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <optional>
+#include <variant>
+
+namespace CUDF_EXPORT cudf {
+
+/**
+ * @addtogroup column_join
+ * @{
+ * @file
+ */
+
+/**
+ * @brief Class that implements sort-merge algorithm for table joins
+ */
+class sort_merge_join {
+ public:
+  /**
+   * @brief Construct a sort-merge join object that pre-processes the right table
+   * on creation, and can be used on subsequent join operations with multiple
+   * left tables.
+   *
+   * @note The `sort_merge_join` object must not outlive the table viewed by `right`,
+   * else behavior is undefined.
+   *
+   * @param right The right table
```

Comment:
I wonder if there is value in referring to the left and right sides of the join as the "stream/probe side" and the "build" side, to use the common parlance.  (That seems to be how they're referred to in #16025.)

Reply:
Ah, true.

Reply:
I might be mistaken, but the terms "build" and "probe" are typically associated with hash joins, where one side of the input, usually the right table, is used to construct the hash table. In that context, "right table" and "build table" are often used interchangeably. However, in the case of sort-merge joins, there's no need to "build" anything in the same sense, as both inputs are (or will be) sorted. To maintain consistency and avoid confusion across different join types, it might be better to update our `hash_join` interfaces to use the terms "left" and "right" tables instead, if needed.